### PR TITLE
ignore empty required value if there is a default value

### PR DESCRIPTION
**Pull request by rianfowler**
#### What this PR does / why we need it:
- empty required fields should not block when there is a default value 

#### Which issue(s) this PR fixes:

Fixes #

https://app.shortcut.com/replicated/story/75137/kots-config-validation-failing-when-the-item-is-required-and-has-default-value-and-or-when-hidden

#### Does this PR introduce a user-facing change?

```release-note
- admin console: fixes an an issue where required fields with default values would not be validated 
```

#### Does this PR require documentation?

https://github.com/replicatedhq/replicated-docs/pull/1094

### CODE REVIEW

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 1]
```diff
@@ -339,22 +340,6 @@ class AppConfig extends Component<Props, State> {
...
-
-    if (hasRequiredFieldError) {
-      this.setState({
-        configGroups: newConfigGroups,
-        showErrorsForEmptyValues: true,
-        showValidationError: true,
-      });
-      return;
-    }
-
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:357]: this was to run the required field validation on save. we're delegating this to the backend

File: web/src/features/AppConfig/components/AppConfig.tsx
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:1]: removing all of the FE required field validation. we're going to rely strictly on the BE for this

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 2]
```diff
@@ -384,6 +369,13 @@ class AppConfig extends Component<Props, State> {
         this.setState({ savingConfig: false });
 
         if (!result.success) {
+          if (result.requiredItems?.length) {
+            this.markRequiredItems(result.requiredItems);
+          }
+          if (result.error) {
+            this.setState({ configError: result.error });
+          }
+
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:378]: this is the old way of showing required items. just restoring it here

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 3]
```diff
@@ -491,49 +483,20 @@ class AppConfig extends Component<Props, State> {
...
       const newGroup = { ...group };
       const configGroupValidationErrors = validationErrors?.find(
         (validationError) => validationError.name === group.name
       );
 
-      if (configGroupValidationErrors || hasRequiredValidationError) {
+      // required errors are handled separately
+      if (group?.items?.find((item) => item.error)) {
+        newGroup.hasError = true;
+      }
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:495]: if there is a required item error in the group then we need to expand that group in the side nav so people can quickly get to the error

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 4]
```diff
@@ -837,7 +798,11 @@ class AppConfig extends Component<Props, State> {
                           return (
                             <a
                               className={`u-fontSize--normal u-lineHeight--normal
-                                ${item.validationError ? "has-error" : ""}
+                                ${
+                                  item.validationError || item.error
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:802]: `item.erorr` is where we store the required item error

File: pkg/handlers/config.go [hunk 5]
```diff
@@ -668,17 +668,6 @@ func (h *Handler) CurrentAppConfig(w http.ResponseWriter, r *http.Request) {
 	currentAppConfigResponse.ConfigGroups = []kotsv1beta1.ConfigGroup{}
 	if renderedConfig != nil {
 		currentAppConfigResponse.ConfigGroups = renderedConfig.Spec.Groups
-		validationErrors, err := configvalidation.ValidateConfigSpec(renderedConfig.Spec)
```
Comment (pavansokkenagaraj) [inline pkg/handlers/config.go:671]: remove validation errors from `/GET` for config

File: pkg/kotsadmconfig/validation/validator.go [hunk 6]
```diff
@@ -49,6 +49,17 @@ func isValidatableConfigItem(item kotsv1beta1.ConfigItem) bool {
 		return false
 	}
 
+	// if item is not required and value is empty, no need to validate
+	if !item.Required && item.Value.String() == "" {
+		return false
+	}
```
Comment (pavansokkenagaraj) [inline pkg/kotsadmconfig/validation/validator.go:55]: moved from `pkg/kotsadmconfig/validation/config.go` so all the logic for `isValidatableConfigItem` is at one place

File: pkg/kotsadmconfig/validation/validator.go [hunk 7]
```diff
@@ -49,6 +49,17 @@ func isValidatableConfigItem(item kotsv1beta1.ConfigItem) bool {
...
+
+	// if item is require and value is empty and default value is not empty,
+	// don't validate the default value (as user has not entered a required value yet and could use the default value)
+	if item.Required && item.Value.String() == "" && item.Default.String() != "" {
+		return false
+	}
```
Comment (pavansokkenagaraj) [inline pkg/kotsadmconfig/validation/validator.go:61]: new logic for the detected edge case,

File: pkg/handlers/config.go [hunk 8]
```diff
@@ -668,17 +668,6 @@ func (h *Handler) CurrentAppConfig(w http.ResponseWriter, r *http.Request) {
...
-		if err != nil {
-			// log the error and continue to return the config, so UI can show the config
-			currentAppConfigResponse.Error = "failed to validate config spec"
-			logger.Error(errors.Wrap(err, currentAppConfigResponse.Error))
-		}
-
-		if len(validationErrors) > 0 {
-			currentAppConfigResponse.ValidationErrors = validationErrors
-			logger.Warnf("Validation errors found for config spec: %v", validationErrors)
-		}
```
Comment (pavansokkenagaraj) [inline pkg/handlers/config.go:681]: remove validation errors from `/GET` for config

File: pkg/kotsadmconfig/validation/config.go [hunk 9]
```diff
@@ -68,12 +68,6 @@ func validateConfigItem(item kotsv1beta1.ConfigItem) (*configtypes.ConfigItemVal
 		return nil, errors.Wrap(err, "failed to get validatable value")
 	}
 
-	// if item is not required and value is empty, no need to validate
-	// if item is required and value is empty, need to validate to return error
-	if validatableValue == "" && !item.Required {
-		return nil, nil
-	}
-
```
Comment (pavansokkenagaraj) [inline pkg/kotsadmconfig/validation/config.go:76]: moved from pkg/kotsadmconfig/validation/config.go so all the logic for isValidatableConfigItem is at one place

File: pkg/kotsadmconfig/validation/validator.go (hunk 6 above)
Comment (sgalsaleh) [reply pkg/kotsadmconfig/validation/validator.go:55]: can't this just be `if item.Value.String() == ""`?

File: pkg/kotsadmconfig/validation/validator.go (hunk 7 above)
Comment (sgalsaleh) [reply pkg/kotsadmconfig/validation/validator.go:61]: not necessary with my above comment

Comment (pavansokkenagaraj) [reply pkg/kotsadmconfig/validation/validator.go:61]: Yes, and makes it simpler

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**